### PR TITLE
:sparkles: Add cache option to filter by field

### PR DESCRIPTION
**Pull request by qinqon**
All instance for a same resources are being cached by
controller-runtime, for some use cases this consumes a lot of memory and
CPU. This change add a option to the cache so resources can be filtered
by field.

### CODE REVIEW

File: pkg/cache/internal/informers_map.go [hunk 1]
```diff
@@ -76,15 +70,10 @@ type MapEntry struct {
 // specificInformersMap create and caches Informers for (runtime.Object, schema.GroupVersionKind) pairs.
 // It uses a standard parameter codec constructed based on the given generated Scheme.
 type specificInformersMap struct {
-	// Scheme maps runtime.Objects to GroupVersionKinds
-	Scheme *runtime.Scheme
```
Comment (vincepri) [inline pkg/cache/internal/informers_map.go:80]: This change seems outside the scope of this PR?

File: pkg/cache/internal/informers_map.go [hunk 2]
```diff
@@ -76,15 +70,10 @@ type MapEntry struct {
...
 
 	// config is used to talk to the apiserver
 	config *rest.Config
 
-	// mapper maps GroupVersionKinds to Resources
-	mapper meta.RESTMapper
-
```
Comment (vincepri) [inline pkg/cache/internal/informers_map.go:87]: Same as above, could we keep these?

File: pkg/cache/internal/informers_map.go [hunk 3]
```diff
@@ -97,11 +86,6 @@ type specificInformersMap struct {
 	// stop is the stop channel to stop informers
 	stop <-chan struct{}
 
-	// resync is the base frequency the informers are resynced
-	// a 10 percent jitter will be added to the resync period between informers
-	// so that all informers will not send list requests simultaneously.
-	resync time.Duration
-
```
Comment (vincepri) [inline pkg/cache/internal/informers_map.go:104]: Same as above, could we keep these?

File: pkg/cache/internal/informers_map.go [hunk 4]
```diff
@@ -117,9 +101,7 @@ type specificInformersMap struct {
 	// unstructured objects.
 	createListWatcher createListWatcherFunc
 
-	// namespace is the namespace that all ListWatches are restricted to
-	// default or empty string means all namespaces
-	namespace string
```
Comment (vincepri) [inline pkg/cache/internal/informers_map.go:122]: Same as above, could we keep these?

File: pkg/cache/cache.go [hunk 5]
```diff
@@ -103,6 +103,10 @@ type Options struct {
 	// Namespace restricts the cache's ListWatch to the desired namespace
 	// Default watches all namespaces
 	Namespace string
+
+	// FieldSelectorByResource restricts the cache's ListWatch to the desired
+	// fields per resource
+	FieldSelectorByResource map[string]string
```
Comment (vincepri) [inline pkg/cache/cache.go:109]: What's the key and value expected to look like here? Would it better to use a GroupKind as key instead?

Comment (vincepri) [reply pkg/cache/cache.go:109]: Or GroupResource?

File: pkg/cache/internal/options.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+// Options are the optional arguments for creating a new InformersMap object
+type Options struct {
+	Scheme                  *runtime.Scheme
+	Mapper                  meta.RESTMapper
+	Resync                  *time.Duration
+	Namespace               string
+	FieldSelectorByResource map[string]string
+}
```
Comment (vincepri) [inline pkg/cache/internal/options.go:17]: Given that this is all internal, how are users supposed to use this new feature?

**Review by vincepri — COMMENTED** [review]
Hi @qinqon, thanks for the contribution. Left a few comments, let's please reduce the PR to the absolute minimum set of changes needed for the requested feature addition. Also this doesn't seem to be a 🐛, but rather a new feature ✨

**Comment by qinqon** [comment]
Remove the commit that encapsulate some internal attributes to an internal Options struct and create a small helper to find the fieldselector on the map using the resource from mapper, @vincepri you can take a look now

File: pkg/cache/internal/informers_map.go (hunk 1 above)
Comment (qinqon) [reply pkg/cache/internal/informers_map.go]: Restored

File: pkg/cache/internal/informers_map.go (hunk 2 above)
Comment (qinqon) [reply pkg/cache/internal/informers_map.go]: Restored

File: pkg/cache/internal/informers_map.go (hunk 3 above)
Comment (qinqon) [reply pkg/cache/internal/informers_map.go]: Done

File: pkg/cache/internal/informers_map.go (hunk 4 above)
Comment (qinqon) [reply pkg/cache/internal/informers_map.go]: Done

File: pkg/cache/cache.go (hunk 5 above)
Comment (qinqon) [reply pkg/cache/cache.go]: Used GroupResource so we can match all the versions.

File: pkg/cache/internal/options.go (hunk 6 above)
Comment (qinqon) [reply pkg/cache/internal/options.go]: Removed the internal Options

File: pkg/cache/cache.go [hunk 7]
```diff
@@ -103,6 +103,10 @@ type Options struct {
 	// Namespace restricts the cache's ListWatch to the desired namespace
 	// Default watches all namespaces
 	Namespace string
+
+	// FieldSelectorByResource restricts the cache's ListWatch to the desired
+	// fields per resource
+	FieldSelectorByResource map[schema.GroupResource]string
```
Comment (alvaroaleman) [inline pkg/cache/cache.go:109]: Please add a sample field selector here and link to the kube docs about field selectors, many ppl are not aware that they only work on a very limited set of fields and not at all for CRDs

**Comment by alvaroaleman** [comment]
@qinqon this doesn't close #244 as it doesn't allow filtering by labels (which I would assume to be more common), so please remove that from the PR body

File: pkg/cache/internal/informers_map.go [hunk 8]
```diff
@@ -231,6 +235,15 @@ func (ip *specificInformersMap) addInformerToMap(gvk schema.GroupVersionKind, ob
 	return i, ip.started, nil
 }
 
+func (ip *specificInformersMap) findFieldSelectorByGVR(gvr schema.GroupVersionResource) string {
+	gr := schema.GroupResource{Group: gvr.Group, Resource: gvr.Resource}
+	fieldSelector, ok := ip.fieldSelectorByResource[gr]
```
Comment (alvaroaleman) [inline pkg/cache/internal/informers_map.go:240]: Nit: This is an empty string if there isn't a match in the map, so you do not need the `ok` check

Comment (qinqon) [reply pkg/cache/internal/informers_map.go]: Done

**Comment by qinqon** [comment]
> @qinqon this doesn't close #244 as it doesn't allow filtering by labels (which I would assume to be more common), so please remove that from the PR body

I can try to add labels filtering at a follow up PR to keep this small.

File: pkg/cache/cache.go (hunk 7 above)
Comment (qinqon) [reply pkg/cache/cache.go:112]: Done

Comment (alvaroaleman) [reply pkg/cache/cache.go:112]: its `medadata.name`, not `meta.name` or not?

**Review by alvaroaleman — COMMENTED** [review]
Looks good except for the doc sample.

> I can try to add labels filtering at a follow up PR to keep this small.

If  you have the time for that, that would be great but I didn't mean to say that you are in any way obligated to that, just that it would be premature to close the issue

File: pkg/cache/cache.go (hunk 7 above)
Comment (qinqon) [reply pkg/cache/cache.go:112]: Done, correct value is `metadata.name`.

**Comment by qinqon** [comment]
@vincepri do you know if Is the issue at pull-controller-runtime-apidiff-master  important ?

**Comment by qinqon** [comment]
[force-push](https://github.com/kubernetes-sigs/controller-runtime/compare/03473b4df63daf398deebb562c0cf2e156461c86..a4e89aedfd10f66d915f4f99d23acf4d0be1aa87): Rebase on master

**Comment by alvaroaleman** [comment]
@qinqon The one thing this change ignores entirely is what impact this has for the cache-backed client. Could you write a small design doc in a different PR that:
* Outlines the motivation for this and possible alternatives
* What this means for the cache-backed client
* Also: Are there ways to defined a fieldSelector other than via a plain string?

Doesn't have to be anything big, but we asked for that multiple times in https://github.com/kubernetes-sigs/controller-runtime/issues/244#issuecomment-451317507 and declined other PRs because of that. I personally like the simplicity of just saying "We allow this on the cache and if you use a custom cache on your manager, you imply that you know better" _but_ its a bigger design question and I want to be sure there is agreement about that.

**Comment by qinqon** [comment]
> @qinqon The one thing this change ignores entirely is what impact this has for the cache-backed client. Could you write a small design doc in a different PR

@alvaroaleman done a PR with the proposal https://github.com/kubernetes-sigs/controller-runtime/pull/1419

**Comment by qinqon** [comment]
[force-push](https://github.com/kubernetes-sigs/controller-runtime/compare/a4e89aedfd10f66d915f4f99d23acf4d0be1aa87..b6217c63f49236118d0d10c4282da2f6f7260aed): Rebased and converting the map value to `fields.Selector`
